sycl/include/CL/sycl/detail/pi.hpp

```diff
@@ -80,13 +80,72 @@ namespace pi {
     template<typename Exception>
     void check(PiResult Result);
   };
+
+  // The run-time tracing of PI calls.
+  // TODO: replace PiCall completely with this one (PiTrace)
+  //
+  template <typename T> inline
+  void Print(T val) {
+    std::cout << "<unknown> : " << val;
+  }
+
+  template<> inline void Print<> (PiPlatform val) { std::cout << "pi_platform : " << val; }
+  template<> inline void Print<> (PiResult val) {
+    std::cout << "pi_result : ";
+    if (val == PI_SUCCESS)
+      std::cout << "PI_SUCCESS";
+    else
+      std::cout << val; 
+  }
+  
+  inline void PrintArgs(void) {}
+  template <typename Arg0, typename... Args>
+  void PrintArgs(Arg0 arg0, Args... args) {
+    std::cout << std::endl << "       ";
+    Print(arg0);
+    PrintArgs(std::forward<Args>(args)...);
+  }
+  
+  template <typename FnType>
+  class Trace {
+  private:
+    FnType m_FnPtr;
+    static bool m_TraceEnabled;
+  public:
+    Trace(FnType FnPtr, const std::string &FnName) : m_FnPtr(FnPtr) {
+      if (m_TraceEnabled)
+        std::cout << "---> " << FnName << "(";
+    }
+  
+    template <typename... Args>
+    typename std::result_of<FnType(Args...)>::type
+    operator() (Args... args) {
+      if (m_TraceEnabled)
+        PrintArgs(args...);
+
+      piInitialize();
+      auto r = m_FnPtr(args...);
+
+      if (m_TraceEnabled) {
+        std::cout << ") ---> ";
+        std::cout << (Print(r),"") << "\n";
+      }
+      return r;
+    }
+  };
+
+  template <typename FnType>
+  bool Trace<FnType>::m_TraceEnabled = (std::getenv("SYCL_PI_TRACE") != nullptr);
```

Comment:
BTW.  Don't think it's problem for tracing, but you will have getenv called for each pi function.

Reply:
Right. It's only once for every (of not so many) PI interfaces, and we can fix this if this is ever going to be a  problem